vowpalwabbit/csoaa.cc

```diff
@@ -462,10 +462,52 @@ void do_actual_learning(ldf& data, single_learner& base, multi_ex& ec_seq_all)
 
   /////////////////////// add headers
   uint32_t K = (uint32_t)ec_seq.size();
+  uint32_t predicted_K = 0;
 
   bool isTest = test_ldf_sequence(data, ec_seq);
+
+  auto restore_guard = VW::scope_exit([&data, &ec_seq, &K, &predicted_K] {
+    if (data.rank)
+    {
+      data.stored_preds[0].clear();
+      for (size_t k = 0; k < K; k++)
+      {
+        ec_seq[k]->pred.a_s = data.stored_preds[k];
+        ec_seq[0]->pred.a_s.push_back(data.a_s[k]);
+      }
+    }
+    else
+    {
+      // Mark the predicted subexample with its class_index, all other with 0
+      for (size_t k = 0; k < K; k++)
+      {
+        if (k == predicted_K)
+          ec_seq[k]->pred.multiclass = ec_seq[k]->l.cs.costs[0].class_index;
```

Comment:
Could we run into the situation that we throw with a label that is not a cost sensitive one?

Reply:
Good point - I have added a restore guard for one of the potential paths (if we throw through `make_single_prediction` line 536)

For the other potential paths `do_actual_learning_wap` and `do_actual_learning_oaa` they already have restore guards for the label